cranelift/codegen/src/machinst/vcode.rs

```diff
@@ -735,3 +757,141 @@ impl<I: VCodeInst> PrettyPrint for VCode<I> {
+    /// will be emitted to the `MachBuffer`. The current implementation can deduplicate constants
+    /// that are [VCodeConstantData::Pool] or [VCodeConstantData::WellKnown] but not
+    /// [VCodeConstantData::Generated].
+    pub fn insert(&mut self, data: VCodeConstantData) -> VCodeConstant {
+        match data {
+            VCodeConstantData::Generated(_) => self.constants.push(data),
+            VCodeConstantData::Pool(constant, _) => match self.pool_uses.get(&constant) {
+                None => {
+                    let vcode_constant = self.constants.push(data);
+                    self.pool_uses.insert(constant, vcode_constant);
+                    vcode_constant
+                }
+                Some(&vcode_constant) => vcode_constant,
+            },
+            VCodeConstantData::WellKnown(data_ref) => {
+                match self.well_known_uses.get(&(data_ref as *const [u8])) {
+                    None => {
+                        let vcode_constant = self.constants.push(data);
+                        self.well_known_uses
+                            .insert(data_ref as *const [u8], vcode_constant);
+                        vcode_constant
+                    }
+                    Some(&vcode_constant) => vcode_constant,
+                }
+            }
+        }
+    }
+
+    /// Retrieve a byte slice for the given [VCodeConstant], if available.
+    pub fn get(&self, constant: VCodeConstant) -> Option<&[u8]> {
+        self.constants.get(constant).map(|d| d.as_slice())
+    }
+
+    /// Return the number of constants inserted.
+    pub fn len(&self) -> usize {
+        self.constants.len()
+    }
+
+    /// Iterate over the [VCodeConstant] keys inserted in this structure.
+    pub fn keys(&self) -> Keys<VCodeConstant> {
+        self.constants.keys()
+    }
+
+    /// Iterate over the [VCodeConstant] keys and the data (as a byte slice) inserted in this
+    /// structure.
+    pub fn iter(&self) -> impl Iterator<Item = (VCodeConstant, &VCodeConstantData)> {
+        self.constants.iter()
+    }
+}
+
+/// A use of a constant by one or more VCode instructions; see [VCodeConstants].
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct VCodeConstant(u32);
+entity_impl!(VCodeConstant);
+
+/// Identify the different types of constant that can be inserted into [VCodeConstants]. Tracking
+/// these separately instead of as raw byte buffers allows us to avoid some duplication.
+pub enum VCodeConstantData {
+    /// A constant already present in the Cranelift IR
+    /// [ConstantPool](crate::ir::constant::ConstantPool).
+    Pool(Constant, ConstantData),
+    /// A reference to a well-known constant value that is statically encoded within the compiler.
+    WellKnown(&'static [u8]),
+    /// A constant value generated during lowering; the value may depend on the instruction context
+    /// which makes it difficult to de-duplicate--if possible, use other variants.
+    Generated(ConstantData),
```

Comment:
@cfallin, I tried adding a `WellKnown` enum to `Amode` but I realized that this way is better: 1) if we want to de-duplicate we can, and 2) all of this constant code is co-located, which makes it a little more comprehensible for other developers (a little of it ends up affecting `MachBuffer` but that is just the easiest way to do things)

Reply:
Hmm, yes, this seems fine; the use-cases for both the static borrow and the inlined/owned data make sense. As a minor design point, folding these two together into a `Cow` might be a bit more idiomatic: we could have e.g. `ConstantDataBytes` that is a type-alias to a `Cow<'static, [u8]>` and then have `VCodeConstantData::Value(ConstantDataBytes)`. Perhaps that's too much type machinery but it would potentially let pass around "owned or borrowed constant bytes" elsewhere.

Reply:
Um, I might not be understanding what you mean but I separated `WellKnown` and `Generated` because `WellKnown` could be `'static` but `Generated` can never be `'static` because the compiler creates it on the fly (e.g. from a shuffle mask). How can I unify them?

Reply:
This is the beauty of [`std::borrow::Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html)! It can hold either a borrow, or an owned thing, and know which is which. So the receiver can get a reference and use the data either way, and the `Cow` takes care of deallocation if and only if it's owned. A `Cow<'static, [u8]>` can either be a `&'static [u8]` or a `Box<[u8]>` under the hood, and either way can give you a `&[u8]` to play with.

Anyway, not a huge deal if we don't use this -- it just stood out to me that the `Generated` and `WellKnown` variants were basically implementing the same pattern here.

Reply:
Ok, I see it now, and I think I could still make the de-duplication work by inspecting into `Cow::Borrowed`. Right now I am leaning towards keeping the current three-way separation because it more clearly informs the user of the decision they are making (we need to explicitly choose a variant and one of those, `Generated`, is not de-duplicated).

Reply:
That's a good point; since this has meaning re: dedup as well I'm fine with the enum as-is. Thanks!